### PR TITLE
rfc: add rfc docs about adding pulsar to tidb-binlog

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <rxl@apache.org>

## Modifications

- Add rfc docs about adding pulsar to tidb-binlog

### CODE REVIEW

**Comment by wolfstudy** [comment]
@winkyao PTAL thanks.

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+Produce the data in source to the topic of Apache Pulsar, sink uses the subscription type of key_shared to consume the data in Topic.
+
+## Compatibility
+
+There is no compatibility issues. Apache Pulsar is compatible with Apache Kafka protocol
+
+## Implementation
+
+1. [@xiaolong](https://github.com/wolfstudy) can support pulsar-client-go, make sure you can meet the required interface in tidb-binlog.
+2. **Tools-Eng** can help to achieve the replacement process
```
Comment (zz-jason) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]: we can use `@pingcap/ecosystem-tools` to replace this part

Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: thanks @zz-jason , will fix it.

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+## Proposal
+
+We will make these changes:
+
+1. Replace **kafak-client-go** with [pulsar-client-go](https://github.com/apache/pulsar-client-go).
+2. Provide a plugin to support more message queues, include: Apahce Pulsar
+
+## Rationale
+
+Produce the data in source to the topic of Apache Pulsar, sink uses the subscription type of key_shared to consume the data in Topic.
```
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]: Sorry I don't quite get these two sentences.

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## Compatibility
+
+There is no compatibility issues. Apache Pulsar is compatible with Apache Kafka protocol
+
+## Implementation
+
+1. [@xiaolong](https://github.com/wolfstudy) can support pulsar-client-go, make sure you can meet the required interface in tidb-binlog.
```
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]: This is supposed to talk about the implementation steps or plans. So i don't quite get it here

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+- Author(s): Xiaolong Ran
+- Last updated:  Dec 05, 2019
+- Discussion at: https://github.com/pingcap/community/issues/117
+
+## Abstract
+
+The proposal aims at:
+
+1. Use **key_shared** subscription type of [Apache Pulsar](https://github.com/apache/pulsar) to enhance **binlog** data processing capabilities;
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: Good idea, will fix it

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 3 above)
Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: yes, I know. But unfortunately, I don't know the implementation details of tidb-binlog here. Maybe it needs to support @pingcap / ecosystem-tools team.

Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: @GregoryIan PTAL in here.

**Comment by wolfstudy** [comment]
@dcalvin fix commends done. PTAL again, thanks.

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+## Abstract
+
+The proposal aims at:
+
+Use **key_shared** subscription type of [Apache Pulsar](https://github.com/apache/pulsar) to enhance **binlog** data processing capabilities;
+
+## Background
+
+During the Change Data Capture(CDC), we need to ensure the order of the messages, or, we need to ensure that DDL arrives before DML. Currently, we use Kafka to process the order of messages, but as you know, Kafka can only guarantee the order of messages within a single partition. If we need to expand the downstream data processing capabilities, sometimes we want more partitions to provide more processing power, but how do we ensure the order of messages at this time? This is a trickier question.
```
Comment (GregoryIan) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]:
```suggestion
During the Change Data Capture(CDC), we need to ensure the order of the messages. And we use Kafka to process the order of messages, but as you know, Kafka can only guarantee the order of messages within a single partition. If we need to expand the downstream data processing capabilities, sometimes we want more partitions to provide more processing power, but how do we ensure the order of messages at this time? This is a trickier question.
```

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+## Proposal
+
+We will make these changes:
+
+1. Replace **kafak-client-go** with [pulsar-client-go](https://github.com/apache/pulsar-client-go).
+2. Provide a plugin to support more message queues, including: Apache Pulsar
```
Comment (GregoryIan) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]: a plugin?  What does it mean? Is there a more detailed explanation?

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+## Rationale
```
Comment (GregoryIan) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]: There are no `Portal` and `Stargate` anymore, I think you can remove this

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 5 above)
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]:
```suggestion
During the Change Data Capture(CDC), we need to ensure the order of the messages, which we use Kafka to process. However, Kafka can only guarantee the order of messages within a single partition. If we need to expand data processing capabilities for downstream, sometimes more partitions are needed. How do we ensure the order of messages in this scenario is a tricky question to tackle.

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+Portal(tidb-binlog source) collects messages from other databases and publishes the messages to Pulsar topics.
+Stargate(tidb-binlog sink) consumes the messages from the Pulsar topics through the key_shared subscription mode and publishes the messages to TiDB.
+## Compatibility
```
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]:
```suggestion

## Compatibility
```

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+There is no compatibility issues. Apache Pulsar is compatible with Apache Kafka protocol
```
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]:
```suggestion
There are no compatibility issues. Apache Pulsar is compatible with the Apache Kafka protocol.
```

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 6 above)
Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: We can make an abstraction of the existing interface implementation, aka, expose an interface, Kafka and Pulsar can implement this interface. 

In this way, all Message Queues that implement this interface can be integrated into tidb-binlog, right?

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 7 above)
Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: Sure, will remove them.

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 6 above)
Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: The rfc's most direct purpose is to replace Kafka with Pulsar. 

In here, I was wondering if we have a more general way to take into account the compatibility of the version and make less intrusion into the original code.

Comment (wolfstudy) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: The rfc's most direct purpose is to replace Kafka with Pulsar. 

In here, I was wondering if we have a more general way to take into account the compatibility of the version and make less intrusion into the original code.

**Comment by wolfstudy** [comment]
@dcalvin @GregoryIan PTAL again, thanks.

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 3 above)
Comment (GregoryIan) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: We have docked this project before 🤔  , but according to the interface of the sarama client, it must meet requirements. But I suggest another better way, define a unified interface for all MQ in golang way, can you help with it @wolfstudy ?

File: rfc/2019-12-10-add-pulsar-to-binlog.md (hunk 6 above)
Comment (GregoryIan) [reply rfc/2019-12-10-add-pulsar-to-binlog.md]: right, same with [my comment above](https://github.com/pingcap/community/pull/136#discussion_r358647539) , just do it

**Comment by GregoryIan** [comment]
LGTM

**Comment by wolfstudy** [comment]
ping @dcalvin PTAL, thanks.

**Comment by wolfstudy** [comment]
ping @dcalvin any update for this?

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+- Author(s): Xiaolong Ran
+- Last updated:  Dec 05, 2019
+- Discussion at: https://github.com/pingcap/community/issues/117
+
+## Abstract
+
+The proposal aims at:
+
+Use **key_shared** subscription type of [Apache Pulsar](https://github.com/apache/pulsar) to enhance **binlog** data processing capabilities;
```
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]: The proposal aims at using the **key_shared** subscription type of [Apache Pulsar](https://github.com/apache/pulsar) to enhance **binlog** data processing capabilities.

File: rfc/2019-12-10-add-pulsar-to-binlog.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+`tidb-binlog` source collects messages from other databases and publishes the messages to Pulsar topics.
+`tidb-binlog` sink consumes the messages from the Pulsar topics through the key_shared subscription mode and publishes the messages to TiDB.
+
+## Compatibility
+
+There are no compatibility issues. Apache Pulsar is compatible with Apache Kafka protocol
+
+## Implementation
+
+1. [@xiaolong](https://github.com/wolfstudy) can support pulsar-client-go, make sure you can meet the required interface in tidb-binlog.
```
Comment (dcalvin) [inline rfc/2019-12-10-add-pulsar-to-binlog.md]:
```suggestion
Implement `pulsar-client-go`, and make sure it meets the required interface in tidb-binlog.
```
If we don't talk about more detailed implementations, the above sentence should be fine.

**Comment by dcalvin** [comment]
> ping @dcalvin any update for this?

Sorry for the late response. Just two suggest edits. Rest LGTM

**Comment by wolfstudy** [comment]
@dcalvin done, please take a look again, thanks.

**Pull request merged**